dataflowengineoss/src/main/scala/io/joern/dataflowengineoss/passes/reachingdef/DdgGenerator.scala

```diff
@@ -31,96 +29,77 @@ class DdgGenerator(semantics: Semantics) {
       */
     def addEdgesToCallSite(call: Call): Unit = {
       // Edges between arguments of call sites
-      usageAnalyzer.usedIncomingDefs(call).foreach { case (use, ins) =>
-        ins.foreach { in =>
-          val inNode = numberToNode(in)
-          if (inNode != use) {
-            addEdge(inNode, use, nodeToEdgeLabel(inNode))
-          }
-        }
-      }
+      call.argument.foreach(addEdgesToNode)
 
       // For all calls, assume that input arguments
       // taint corresponding output arguments
       // and the return value. We filter invalid
       // edges at query time (according to the given semantic).
-      usageAnalyzer.uses(call).foreach { use =>
-        gen(call).foreach { g =>
-          val genNode = numberToNode(g)
-          if (use != genNode && isDdgNode(use)) {
-            addEdge(use, genNode, nodeToEdgeLabel(use))
-          }
-        }
-      }
+      for {
+        use <- call.argument.toList
+        g   <- gen(call).toList
+        genNode = numberToNode(g)
+        if use != genNode && isDdgNode(use)
+      } addEdge(use, genNode, nodeToEdgeLabel(use))
 
       // Handle block arguments
       // This handles `foo(new Bar())`, which is lowered to
       // `foo({Bar tmp = Bar.alloc(); tmp.init(); tmp})`
-      call.argument.isBlock.foreach { block =>
-        block.astChildren.lastOption match {
-          case None => // Do nothing
-          case Some(node: Identifier) =>
-            val edgesToAdd = in(node).toList.flatMap { inDef =>
-              numberToNode.get(inDef) match {
-                case Some(identifier: Identifier) => Some(identifier)
-                case Some(call: Call)             => Some(call)
-                case _                            => None
-              }
-            }
-            edgesToAdd.foreach { inNode =>
-              addEdge(inNode, block, nodeToEdgeLabel(inNode))
-            }
-            if (edgesToAdd.nonEmpty) {
-              addEdge(block, call)
-            }
-          case Some(node: Call) =>
-            addEdge(node, call, nodeToEdgeLabel(node))
-            addEdge(block, call)
-          case _ => // Do nothing
-        }
+      for {
+        block <- call.argument.isBlock
+        last  <- block.astChildren.lastOption
+      } {
+        addEdgesToNode(last)
+        addEdge(last, block, nodeToEdgeLabel(last))
+        addEdge(block, call)
       }
 
     }
 
     def addEdgesToReturn(ret: Return): Unit = {
-      usageAnalyzer.usedIncomingDefs(ret).foreach { case (use: CfgNode, inElements) =>
-        addEdge(use, ret, use.code)
-        inElements
-          .filterNot(x => numberToNode.get(x).contains(use))
-          .flatMap(numberToNode.get)
-          .foreach { inElemNode =>
-            addEdge(inElemNode, ret, nodeToEdgeLabel(inElemNode))
-          }
-        if (inElements.isEmpty) {
-          addEdge(method, ret)
-        }
-      }
+      val uses = ret.astChildren.collect { case x: Expression => x }.toList
```

Comment:
```suggestion
      val uses = ret.astChildren.collectAll[Expression].toList
```

Reply:
Looks like a lot of tests are broken. I will try fix them